### PR TITLE
Fix mp empty side1

### LINKED ISSUE #2563: OOS in multiplayer when [modify_side] changes a controller during "side_turn_end" event.
if a code like 
```
[modify_side]
  side = $($current_side + 1)
  controller=human
[/modify_side]
```
is executed during an "side_turn_end" OOS can happen (assumng that side was previously null-controlled)
What happens is that the [end_turn] command is sended to the server before that event is executed and the server updates the current_player_ field when it recieves [end_turn]. So what happens on the server is that a the time when current_player_ filed is updated the next side is skipped becasue it is empty, on the clients however current_side_ is updated after the 'side_turn_end' event has finished so the next side will not be skipped since modify_side activated it.

### PR DESCRIPTION
**Pull request by gfgtdf**
We refactor a the mp server game turn data a bit to fix #2563 and #3899

### CODE REVIEW

File: src/server/wesnothd/game.hpp [hunk 1]
```diff
@@ -483,6 +481,24 @@ class game
 		return description_;
 	}
 
+	simple_wml::node* description_for_writing()
```
Comment (stevecotton) [inline src/server/wesnothd/game.hpp:484]: Needs documentation

File: src/server/wesnothd/game.hpp [hunk 2]
```diff
@@ -749,7 +766,8 @@ class game
 	 * @param new_side The side number whose turn to move it has become.
 	 * @return True if the current side and-or current turn values have been updated, false otherwise.
```
Comment (stevecotton) [inline src/server/wesnothd/game.hpp:767]: Docs out of date

File: src/server/wesnothd/game.hpp [hunk 3]
```diff
@@ -870,10 +895,18 @@ class game
 	/** Pointer to the game's description in the games_and_users_list_. */
 	simple_wml::node* description_;
 
+	/** Set to true whene ever description_ was changed so that it needs to be sent and update to the clients */
```
Comment (stevecotton) [inline src/server/wesnothd/game.hpp:898]: Typo of "whenever"

File: src/server/wesnothd/game.hpp [hunk 4]
```diff
@@ -870,10 +895,18 @@ class game
...
+	bool description_updated_;
+
 	/** The game's current turn. */
 	int current_turn_;
 	/** The index of the current side. The side number is current_side_index_+1. */
 	int current_side_index_;
+	/**
+	 * after [end_turn] was received, this contains the side for who we accept [init_side].
+	 * -1 if we currently dont accept [init_side] because the current player didn't end his turn yet.
```
Comment (stevecotton) [inline src/server/wesnothd/game.hpp:907]: Typo of "don't"

File: src/server/wesnothd/game.cpp [hunk 5]
```diff
@@ -907,6 +913,14 @@ bool game::is_legal_command(const simple_wml::node& command, player_iterator use
 		}
 	}
 
+	if(command.child("init_side")) {
+		if(next_side_index_ != -1 && sides_[get_next_side_index()] == user) {
+			return true;
```
[the remark this answers is not in the document]
Comment (gfgtdf) [reply src/server/wesnothd/game.cpp:918]: no that's exactly what the `sides_[get_next_side_index()] == user` does, `user` is the user who send this, and  `sides_[get_next_side_index()] ` is the user who is allowed to send this.

**Comment by Pentarctagon** [comment]
@gfgtdf please address the last remaining review comment, and then I think this can be merged?

**Review by soliton- — CHANGES_REQUESTED** [review]

File: src/server/wesnothd/server.cpp [hunk 6]
```diff
@@ -1919,11 +1918,12 @@ void server::handle_player_in_game(player_iterator p, simple_wml::document& data
 		data.root().has_attr("side")
 	) {
 		return;
-	}
-
-	WRN_SERVER << p->client_ip() << "\tReceived unknown data from: " << player.name()
+	} else {
```
[the remark this answers is not in the document]
Comment (gfgtdf) [reply src/server/wesnothd/server.cpp:1921]: All of these end with a return, having this outside make its looks like not handled with a return is a possibility.

File: src/server/wesnothd/game.cpp [hunk 7]
```diff
@@ -1998,4 +2001,26 @@ bool game::is_reload() const
...
+	return get_next_nonempty(next_side_index_);
+}
+
+int game::get_next_nonempty(int side_index) const
+{
+	if(side_index == -1) {
+		return -1;
+	}
+	if(nsides_ == 0) {
+		return 0;
```
[the remark this answers is not in the document]
Comment (gfgtdf) [reply src/server/wesnothd/game.cpp:2015]: this is mainly to match the previous behvaior of current_side() which returned 0 in that case and now uses this function.

File: src/server/wesnothd/server.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (soliton-) [reply src/server/wesnothd/server.cpp:1921]: Of course it's a possibility and it does happen. Are you suggesting that to anyone reading the code the final warning suggests not handling a case is wanted? And if the final warning is in an else case then it suggest you must handle all cases?

[the remark this answers is not in the document]
Comment (gfgtdf) [reply src/server/wesnothd/server.cpp:1921]: where does it happen? i didn't find anything

**Pull request merged**